std/traits.d

```diff
@@ -8061,3 +8062,2432 @@ enum isCopyable(S) = is(typeof(
+{
+    struct Bar
+    {
+        string bar1(double d) @safe { return ""; }
+    }
+
+    Bar bar;
+    string z1 = bar.bar1(1);
+
+    struct Baz
+    {
+        string bar1(double d) @safe { return ""; }
+    }
+
+    Baz baz;
+    string z2 = baz.bar1(1);
+
+    static assert(areTypesEqual!(Baz.bar1, Bar.bar1));
+    static assert(areTypesEqual!(Bar.bar1, Baz.bar1));
+}
+
+@system pure nothrow @nogc
+unittest
+{
+    struct Bar
+    {
+        string bar1(double d) @system pure nothrow @nogc { return ""; }
+    }
+
+    Bar bar;
+    string z1 = bar.bar1(1);
+
+    struct Baz
+    {
+        string bar1(double d) @safe pure nothrow @nogc { return ""; }
+    }
+
+    Baz baz;
+    string z2 = baz.bar1(1);
+
+    static assert(!areTypesEqual!(Baz.bar1, Bar.bar1));
+    static assert(!areTypesEqual!(Bar.bar1, Baz.bar1));
+}
+
+@safe pure nothrow @nogc
+unittest
+{
+    struct Bar
+    {
+        int bar1() @safe pure nothrow @nogc { return 0; }
+        void bar2(int x) @safe pure nothrow @nogc { }
+    }
+
+    Bar bar;
+    int z1 = bar.bar1();
+    bar.bar2(1);
+
+    struct Baz
+    {
+        int bar1() @safe pure nothrow @nogc { return 0; }
+        void bar2(int x) @safe pure nothrow @nogc { }
+    }
+
+    Baz baz;
+    int z2 = baz.bar1();
+    baz.bar2(1);
+
+    static assert(areTypesEqual!(Baz.bar1, Bar.bar1));
+    static assert(areTypesEqual!(Bar.bar1, Baz.bar1));
+    static assert(areTypesEqual!(Baz.bar2, Bar.bar2));
+    static assert(areTypesEqual!(Bar.bar2, Baz.bar2));
+}
+
+@safe pure nothrow @nogc
+unittest
+{
+    struct Bar
+    {
+        auto bar1(int x) { return x; }
+    }
+
+    Bar bar;
+    int z1 = bar.bar1(1);
+
+    struct Baz
+    {
+        auto bar1(int x) { return x; }
+    }
+
+    Baz baz;
+    int z2 = baz.bar1(1);
+
+    static assert(areTypesEqual!(Baz.bar1, Bar.bar1));
+    static assert(areTypesEqual!(Bar.bar1, Baz.bar1));
+}
+
+@safe pure nothrow @nogc
+unittest
+{
+    struct Bar
+    {
+        auto bar1(T)(T x) { return x; }
+    }
+
+    Bar bar;
+    int z1 = bar.bar1(1);
+
+    struct Baz
+    {
+        auto bar1(T)(T x) { return x; }
+    }
+
+    Baz baz;
+    int z2 = baz.bar1(1);
+
+    static assert(areTypesEqual!(Baz.bar1, Bar.bar1));
+    static assert(areTypesEqual!(Bar.bar1, Baz.bar1));
+}
+
+/// Returns true if `T` has alias this member.
+private template hasAliasThis(T)
+{
+    import std.meta : AliasSeq;
+
+    alias members = AliasSeq!(__traits(getAliasThis, T));
+
+    static if (members.length == 0)
+    {
+        enum bool hasAliasThis = false;
+    }
+    else
+    {
+        enum bool hasAliasThis = true;
+    }
+}
+
+@safe pure nothrow @nogc
+unittest
+{
+    struct Foo
+    {
+    }
+
+    struct Bar
+    {
+        Foo fooMemb;
+        alias fooMemb this;
+    }
+
+    static assert(!hasAliasThis!Foo);
+    static assert(hasAliasThis!Bar);
+}
+
+/// DMD does not support multiple alias this, but spec says it may in future.
+private template hasMultipleAliasThis(T)
```

Comment:
No need to support features that don't exist, and might never exist.

Reply:
@JackStouffer I disagree. The D language spec under 15.18
https://dlang.org/spec/class.html#alias-this
specifically says that multiple alias this is allowed, but currently unimplemented. I included the template on this basis. If the feature may never exist, the spec should be adjusted to reflect this. I wrote the functionality based on the spec. It is relatively trivial to include multiple alias this tests in the future, but I wanted to provide some sort of error in case the feature was added, rather than have it silently ignore the additional alias this's.